pkg/rhcos/builds.go

```diff
@@ -5,69 +5,66 @@ import (
-	} `json:"amis"`
-	Azure struct {
-		Image string `json:"image"`
-		URL   string `json:"url"`
-	}
-	GCP struct {
-		Image   string `json:"image"`
-		Project string `json:"project"`
-		URL     string `json:"url"`
-	}
-	BaseURI string `json:"baseURI"`
-	Images  struct {
-		QEMU struct {
-			Path               string `json:"path"`
-			SHA256             string `json:"sha256"`
-			UncompressedSHA256 string `json:"uncompressed-sha256"`
-		} `json:"qemu"`
-		OpenStack struct {
-			Path               string `json:"path"`
-			SHA256             string `json:"sha256"`
-			UncompressedSHA256 string `json:"uncompressed-sha256"`
-		} `json:"openstack"`
-		VMware struct {
-			Path   string `json:"path"`
-			SHA256 string `json:"sha256"`
-		} `json:"vmware"`
-	} `json:"images"`
-	OSTreeVersion string `json:"ostree-version"`
-}
-
-func fetchRHCOSBuild(ctx context.Context, arch types.Architecture) (*metadata, error) {
-	file, err := data.Assets.Open(fmt.Sprintf("rhcos-%s.json", arch))
+// FetchCoreOSBuild returns the pinned version of RHEL/Fedora CoreOS used
+// by the installer to provision the bootstrap node and control plane currently.
+// For more information, see e.g. https://github.com/openshift/enhancements/pull/201
+func FetchCoreOSBuild(ctx context.Context) (*stream.Stream, error) {
+	filename := fmt.Sprintf("%s.json", pinnedRHCOSStream)
+	file, err := data.Assets.Open(filename)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "failed to read embedded CoreOS stream metadata %s", filename)
 	}
 	defer file.Close()
 
 	body, err := ioutil.ReadAll(file)
-	if os.IsNotExist(err) {
-		return nil, errInvalidArch
-	} else if err != nil {
-		return nil, err
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read CoreOS stream metadata")
 	}
 
-	var meta *metadata
-	if err := json.Unmarshal(body, &meta); err != nil {
-		return meta, errors.Wrap(err, "failed to parse RHCOS build metadata")
+	var st stream.Stream
+	if err := json.Unmarshal(body, &st); err != nil {
+		return nil, errors.Wrap(err, "failed to parse CoreOS stream metadata")
 	}
+	return &st, nil
+}
+
+// FormatURLWithIntegrity squashes an artifact into a URL string
+// with the uncompressed sha256 as a query parameter.  This is necessary
+// currently because various parts of the installer pass around this
+// reference as a string, and it's also exposed to users via install-config overrides.
+func FormatURLWithIntegrity(artifact *stream.Artifact) (string, error) {
+	u, err := url.Parse(artifact.Location)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse artifact URL: %v", err)
+	}
+	q := u.Query()
+	q.Set("sha256", artifact.UncompressedSha256)
+	u.RawQuery = q.Encode()
+	return u.String(), nil
+}
 
-	return meta, nil
+// FindArtifactURL uses the first recognized artifact type; this
+// mainly abstracts over e.g. `qcow2.xz` and `qcow2.gz`.  The main
+// platform with multiple artifact types today is `metal`; don't
+// use this for that platform.
```

Comment:
I don't understand the last sentence. The way that I read it is that this function is mainly to accommodate the `metal` platform, but also don't use this function for the `metal` platform.